File: benches/fixtures/medium.js
// Medium-sized JavaScript bench fixture: exercises classes, arrow
// fns, destructuring, template literals, async/await, and spread.
// Self-contained so the grammar sees the full surface without an
// import cluster obscuring it.

const DEFAULTS = Object.freeze({ size: 16, verbose: false });

class Counter {
  constructor(name, options = {}) {
    this.name = name;
    this.options = { ...DEFAULTS, ...options };
    this.values = [];
    this.histogram = new Map();
  }

  push(v) {
    const key = typeof v === "string" ? v : String(v);
    this.histogram.set(key, (this.histogram.get(key) ?? 0) + 1);
    this.values.push(v);
    return this;
  }

  get size() {
    return this.values.length;
  }

  *entries() {
    for (const [k, n] of this.histogram) {
      yield { key: k, count: n };
    }
  }

  summary() {
    const top = [...this.histogram.entries()]
      .sort(([, a], [, b]) => b - a)
      .slice(0, 3)
      .map(([k, n]) => `${k}=${n}`)
      .join(", ");
    return `${this.name} (n=${this.size}, top=[${top}])`;
  }
}

function classify(n) {
  switch (true) {
    case n < 0:
      return "negative";
    case n === 0:
      return "zero";
    case n < 10:
      return "single";
    case n < 100:
      return "double";
    default:
      return "large";
  }
}

async function loadSamples(source) {
  try {
    const data = await source.read();
    return data.filter(x => x != null).map(x => Number(x));
  } catch (e) {
    console.error(`load failed: ${e.message}`);
    return [];
  } finally {
    source.close?.();
  }
}

const sink = {
  log(msg) {
    console.log(msg);
  },
  warn(msg) {
    console.warn(`WARN: ${msg}`);
  },
};

const data = [1, 2, 2, 3, 10, 10, 10, -1];
const c = new Counter("ages");
for (const x of data) c.push(x);
sink.log(c.summary());
for (const n of data) sink.log(`${n} -> ${classify(n)}`);

File: benches/fixtures/medium.rs
//! Medium-sized Rust bench fixture: exercises items, generics,
//! match, closures, and macros. Intentionally self-contained (no
//! external crate imports) so the grammar sees the full surface
//! without a `use` cluster obscuring it.

use std::collections::HashMap;
use std::fmt::{self, Debug};

#[derive(Debug, Clone)]
pub struct Counter<'a, T: Clone + 'a> {
    name: &'a str,
    values: Vec<T>,
    histogram: HashMap<String, usize>,
}

impl<'a, T: Clone + Debug + 'a> Counter<'a, T> {
    pub fn new(name: &'a str) -> Self {
        Self {
            name,
            values: Vec::new(),
            histogram: HashMap::new(),
        }
    }

    pub fn push(&mut self, v: T) {
        let key = format!("{:?}", v);
        *self.histogram.entry(key).or_insert(0) += 1;
        self.values.push(v);
    }

    pub fn summary(&self) -> String {
        let top = self
            .histogram
            .iter()
            .max_by_key(|(_, &n)| n)
            .map(|(k, n)| format!("{}={}", k, n))
            .unwrap_or_else(|| "<empty>".into());
        format!("{} (n={}, top={})", self.name, self.values.len(), top)
    }
}

pub fn classify(n: i64) -> &'static str {
    match n {
        i if i < 0 => "negative",
        0 => "zero",
        1..=9 => "single-digit",
        10..=99 => "two-digit",
        _ => "large",
    }
}

pub fn with_each<F: FnMut(i64)>(xs: &[i64], mut f: F) {
    for &x in xs {
        f(x);
    }
}

fn main() {
    let mut c: Counter<i64> = Counter::new("ages");
    let data = vec![1i64, 2, 2, 3, 10, 10, 10, -1];
    with_each(&data, |x| c.push(x));
    println!("{}", c.summary());
    for n in data {
        println!("{} -> {}", n, classify(n));
    }
}
